oidc/client.py

```diff
@@ -105,6 +106,93 @@ def _retrieve_jwks_related_to_kid(self, kid: str) -> Optional[str]:
                     return jwk
         return None
 
+    def retrieve_user_info(self, access_token: str) -> dict:
+        """
+        Given an access_token, retrieve user profile claims
+        """
+        response = requests.get(
+            self.user_info_endpoint, headers={"Authorization": f"Bearer {access_token}"}
+        )
+        return response.json()
+
+    def get_hash_algorithm(self, alg: str) -> Optional[Callable]:
+        """
+        Maps JWT algorithm to hash function.
+
+        Based on the spec: RS256/ES256/PS256 use SHA-256,
+        RS384/ES384/PS384 use SHA-384, RS512/ES512/PS512 use SHA-512
+        """
+        algorithm_map = {
+            "RS256": hashlib.sha256,
+            "RS384": hashlib.sha384,
+            "RS512": hashlib.sha512,
+            "ES256": hashlib.sha256,
+            "ES384": hashlib.sha384,
+            "ES512": hashlib.sha512,
+            "PS256": hashlib.sha256,
+            "PS384": hashlib.sha384,
+            "PS512": hashlib.sha512,
+            "HS256": hashlib.sha256,
+            "HS384": hashlib.sha384,
+            "HS512": hashlib.sha512,
+        }
+        return algorithm_map.get(alg)
+
+    def validate_access_token(
+        self, verifyied_id_token: dict, id_token: str, access_token: str
+    ) -> bool:
+        """
+        Validates an access token against the at_hash claim in an ID token.
+
+        Args:
+            verified_id_token: A verified and decoded ID token
+            id_token: The ID token (JWT) as a string
+            access_token: The access token to validate
+
+        Returns:
+            bool: True if valid, False otherwise
+        """
+
+        if "at_hash" not in verifyied_id_token:
+            return False
+
+        id_token_header = jwt.get_unverified_header(id_token)
+        alg_str = id_token_header.get("alg")
+        hash_algorithm = self.get_hash_algorithm(alg_str)
+        if not hash_algorithm:
+            raise ValueError(f"Unsupported algorithm: {alg_str}")
+
+        hash_digest = hash_algorithm(access_token.encode("ascii")).digest()
+        left_half = hash_digest[: len(hash_digest) // 2]
+        computed_at_hash = (
+            base64.urlsafe_b64encode(left_half).decode("ascii").rstrip("=")
+        )
+
+        return computed_at_hash == verifyied_id_token["at_hash"]
+
+    def tokens_to_user_info(
+        self, decoded_id_token: dict, id_token: str, access_token: str
+    ) -> dict:
+        # TODO: need to use a setting for this
+        if (
+            "emails" in decoded_id_token
+            and decoded_id_token["emails"]
+            and decoded_id_token["emails"][0]
+        ):
+            return decoded_id_token
+        if "email" in decoded_id_token:
+            return decoded_id_token
+        elif access_token and self.validate_access_token(
```

Comment:
According to the spec, validation of the access token can happen if the ID token contains `at_hash`. Therefore, we only ought to validate access token if this field is present

> if the ID Token contains an at_hash Claim, the Client MAY use it to validate the Access Token